js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,196 @@
+  protected params: any = {};
+
+  protected initialLoading: boolean = false;
+  protected loadingPrev: boolean = false;
+  protected loadingNext: boolean = false;
+
+  hasPrev: boolean = false;
+  hasNext: boolean = false;
+
+  protected constructor(offset: number = 20) {
+    this.offset = offset;
+  }
+
+  abstract get type(): string;
+
+  protected parseResults(pageNum, results: T[]) {
+    const page = {
+      number: pageNum,
+      items: results,
+    };
+
+    if (this.isEmpty() || pageNum > this.getNextPageNumber() - 1) {
+      this.pages.push(page);
+    } else {
+      this.pages.unshift(page);
+    }
+
+    this.hasNext = !!results.payload?.links?.next;
+    this.hasPrev = !!results.payload?.links?.prev;
+
+    m.redraw();
+  }
+
+  /**
+   * Load a new page of results.
+   */
+  protected loadPage(page = 1): Promise<T[]> {
+    const params = this.requestParams();
+    params.page = { offset: this.offset * (page - 1) };
+
+    if (Array.isArray(params.include)) {
+      params.include = params.include.join(',');
+    }
+
+    return app.store.find(this.type, params);
+  }
+
+  public hasItems(): boolean {
+    return !!this.getAllItems().length;
+  }
+
+  public isEmpty(): boolean {
+    return !this.isInitialLoading() && !this.hasItems();
+  }
+
+  public getLocation(): PaginationLocation {
+    return this.location;
+  }
+
+  protected getAllItems(): T[] {
+    return this.pages.map((pg) => pg.items).flat();
+  }
+
+  public isInitialLoading(): boolean {
+    return this.initialLoading;
+  }
+  public isLoadingPrev(): boolean {
+    return this.loadingPrev;
+  }
+  public isLoadingNext(): boolean {
+    return this.loadingNext;
+  }
+
+  /**
+   * Get the parameters that should be passed in the API request.
+   * Do not include page offset unless subclass overrides loadPage.
+   *
+   * @abstract
+   * @see loadPage
+   */
+  protected requestParams(): any {
+    return {};
+  }
+
+  /**
+   * Stored state parameters.
+   */
+  public getParams(): any {
+    return this.params;
+  }
+
+  /**
+   * Update the `this.params` object, calling `refresh` if they have changed.
+   * Use `requestParams` for converting `this.params` into API parameters
+   *
+   * @param newParams
+   * @see requestParams
+   */
+  public refreshParams(newParams) {
+    if (this.isEmpty() || Object.keys(newParams).some((key) => this.requestParams()[key] !== newParams[key])) {
+      this.params = newParams;
+
+      return this.refresh();
+    }
+  }
+
+  public refresh(page: number = 1) {
+    this.initialLoading = true;
+    this.loadingPrev = false;
+    this.loadingNext = false;
+
+    m.redraw();
+
+    this.location = { page };
+
+    return this.loadPage()
+      .then((results: T[]) => {
+        this.pages = [];
+        this.parseResults(this.location.page, results);
+      })
+      .finally(() => (this.initialLoading = false));
+  }
+
+  public clear() {
+    this.pages = [];
+    m.redraw();
+  }
+
+  public loadMore() {
```

Comment:
Let's annotate these as async

Reply:
That doesn't really make any difference though? I thought `async` was to use `await` in the body, not do denote any method returning a Promise... not sure if the build dist would change after adding that.